src/main/kotlin/org/danilopianini/plagiarismdetector/analyzer/representation/token/GramImpl.kt

```diff
@@ -0,0 +1,24 @@
+package org.danilopianini.plagiarismdetector.analyzer.representation.token
+
+/**
+ * A simple class implementing a [Gram].
+ */
+data class GramImpl<T>(override val items: Sequence<T>) : Gram<T> {
+    override fun hashCode(): Int {
+        return items.hashCode()
+    }
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) {
+            return true
+        } else if (javaClass != other?.javaClass) {
+            return false
+        }
+        other as GramImpl<*>
+        return other.items.toList() == items.toList()
+    }
+
+    override fun toString(): String {
+        return items.toList().toString()
+    }
+}
```

Comment:
```suggestion
data class GramImpl<T>(override val items: Sequence<T>) : Gram<T> {
    override fun toString(): String = "Gram$items"
}
```

much simpler :)

Reply:
During the development, however, [this](https://github.com/tassiLuca/code-plagiarism-detector/blob/master/src/test/kotlin/org/danilopianini/plagiarismdetector/analyzer/TokenizedSourceTest.kt) test failed as you can see [here](https://github.com/tassiLuca/code-plagiarism-detector/actions/runs/3072878148/jobs/4964648487#step:3:271) (line 259). This is due to the fact in the constructor `items` is typed as a `Sequence`. So I override `equals` and `hashCode`. Would it be better to change it in an `Iterable`?

Reply:
If it is an ordered collection of items, the data structure to use in Kotlin is `List`. `Sequence`s are meant as "pull" pipelines (similar to Java's `Stream`s), to be used when the element count is unknown and the processing should be depth-first. `Flow`s are "push" pipelines.

You can still provide a secondary constructor with a sequence, if needed.